### ISSUE TITLE
java.lang.NoSuchMethodError when adding Chucker with compileOnly in library 

### ISSUE THREAD

**Issue by degill** [ChuckerTeam/chucker#357]
## :writing_hand: Describe the bug
When I add the regular version of Chucker as a `compileOnly` dependency in a library module and have my application module depend on the no-op version via `implementation` then the app crashes when trying to instantiate `ChuckerInterceptor`.

## :bomb: Steps to reproduce

1. Create a library module called `:cloud` and depend on Chucker with `compileOnly "com.github.ChuckerTeam.Chucker:library:3.2.0"`
2. Within `cloud` create an Object of `ChuckerInterceptor`
2. Create an application module `:app`, depend on `:cloud` and depend on Chucker with `implementation "com.github.ChuckerTeam.Chucker:library-no-op:3.2.0"`
3. Start the app

## :wrench: Expected behavior
The app can create an instance of the no-op version of `ChuckerInterceptor`.

## Error message
> java.lang.NoSuchMethodError: No direct method <init>(Landroid/content/Context;Lcom/chuckerteam/chucker/api/ChuckerCollector;JLjava/util/Set;ILkotlin/jvm/internal/DefaultConstructorMarker;)V in class Lcom/chuckerteam/chucker/api/ChuckerInterceptor; or its super classes (declaration of 'com.chuckerteam.chucker.api.ChuckerInterceptor' appears in /data/app/*********==/base.apk)

## :iphone: Tech info
- Kotlin version: 1.3.72
- Chucker version: 3.2.0

## :page_facing_up: Additional context

This only happens when the versions applied in `:app` and `c:loud` differ, that is when one uses the no-op version and the other uses the regular one.

In the `cloud` module I can not decide which version of Chucker I can use. The version is decided for different flavours of the application and those are only available in `:app`. 

Since I do the same with `implementation com.jakewharton.threetenabp:threetenabp` and `compileOnly org.threeten:threetenbp:$threetenbpVersion` I thought it might be possible to provide the final artifact at a different place (i.e. `:app`).

**Comment by cortinico** [comment]
tl;dr: you can't use the `library` and `library-no-op` of Chucker and do dependency shading with `compileOnly`.

The reason behind this is that the two artifacts are not binary compatible each other (that's by design).

The signature of the `ChuckerInterceptor` class is different between the two artifacts:

** library ** 

https://github.com/ChuckerTeam/chucker/blob/583e395c1a94cbc7ea4ce3226161e6bf4ac4d40b/library/src/main/java/com/chuckerteam/chucker/api/ChuckerInterceptor.kt#L39-
L45

** library-no-op ** 

https://github.com/ChuckerTeam/chucker/blob/583e395c1a94cbc7ea4ce3226161e6bf4ac4d40b/library-no-op/src/main/java/com/chuckerteam/chucker/api/ChuckerInterceptor.kt#L11-L16

(See Change type of a formal parameter [here](https://wiki.eclipse.org/Evolving_Java-based_APIs_2#Evolving_API_classes_-_API_methods_and_constructors))

So your app is failing when you're using `library-no-op` as `compileOnly` and `library` as `implementation` in `app` (or viceversa) as one of the two constructor is missing.

The `library-no-op` artifacts is source compatible with `library`, and not vice versa. You can swap `library` with `-no-op` and the code will still compile. 
This allows to compile/assemble on the `release` variant without compilation failures.

Long story short: you need to use `debugImplementation`/`releaseImplementation`. 
I'm closing here. 

Anyway @degill, I'm curious to hear what's the use case behind using `compileOnly`

**Comment by degill** [comment]
Hey @cortinico thank you very much for clarifying! I had the same intuition when I looked at the source code for both classes, but I didn't knew if this would really result in "binary incompatibility". 

My use case is the following: 
I'm building an application that is heavily modularised. So I have one module `:cloud` that's responsible for handling calls to the interwebs and that module is also using Chucker. But my application module `:app` is heavily customisable in the sense that I have all kinds of flavours to build the app. There is one for us internal developers, two or more for our partners (developers and non-developers, designers, etc.) and then finally the store version.

So the flavour dimensions are not only `debug`/`release` but also `develop`/`partnerDevelop`/`partnerDesign`/`store`. And for `develop` and `partnerDevelop` I want to have Chucker enabled, but not for the rest of the flavours. These flavours are not available in my library modules. So only in `:app` I can do something like `partnerDevelopImplementation "com.github.ChuckerTeam.Chucker:library:3.2.0"`.

I think I can workaround this by passing in some build info so I can decide in `:cloud` whether to use Chucker or not.

**Comment by cortinico** [comment]
> So the flavour dimensions are not only debug/release

Please note that `debug/release` are not flavour but build types. Those should be available on your android library modules (also in the `cloud` module). So you should be fine using `debugImplementation` and `releaseImplementation` directly inside cloud.

**Comment by degill** [comment]
Okay sorry, I should've clarified that part. 

Yes, of course `debug/release` are available. But that info is not enough. There will never be a `partnerDesignStoreRelease` version. The only `release` build thats ever going to be made is of the `store` flavour. 

So the usual `debug -> library` and `release -> no-op library` does not work, because even with `debug` I sometimes want the no-op library.
